pkg/sentry/client.go

```diff
@@ -37,12 +41,41 @@ func NewHTTPClient(d doer, pluginId string, b BuildInfoProvider, authToken strin
 
 // Do attaches the sentry authentication header and the User-Agent header to
 // the request and passes it to the injected http Doer
-func (a HTTPClient) Do(req *http.Request) (*http.Response, error) {
+func (a HTTPClient) Do(req *http.Request, rateLimitsRetryOnFailure bool) (*http.Response, error) {
 	req.Header.Set("User-Agent", fmt.Sprintf("%s/%s", a.pluginId, a.version))
 	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", a.authToken))
 	_, E2E_GRAFANA_VERSION_PRESENT := os.LookupEnv("E2E_GRAFANA_VERSION")
 	if E2E_GRAFANA_VERSION_PRESENT {
 		return mocker.Mock(req)
 	}
-	return a.doer.Do(req)
+	retryCount := 0
+	for {
+		resp, err := a.doer.Do(req)
+		if err != nil {
+			return nil, err
+		}
+
+		if resp.StatusCode == http.StatusTooManyRequests && rateLimitsRetryOnFailure && retryCount < maxRetryAttempts {
+			// Check for "X-Sentry-Rate-Limit-Reset" header
+			resetTimeStr := resp.Header.Get("X-Sentry-Rate-Limit-Reset")
+			if resetTimeStr != "" {
+				resetTime, err := strconv.ParseInt(resetTimeStr, 10, 64)
+				if err != nil {
+					return nil, err
+				}
+				currentTime := time.Now().Unix()
+
+				if resetTime > currentTime {
```

Comment:
How long can the rate limit reset be? It may be the case that the user's query times out on the Grafana server before this loop can complete